examples/ctc.dx

```diff
@@ -0,0 +1,134 @@
+'## Connectionist Temporal Classification
+By Alex Graves et alia.
+
+'[Link to paper](https://www.cs.toronto.edu/~graves/icml_2006.pdf).
+
+' This dynamic programming algorithm computes the probability of a particular sequences of labels
+(without pauses included) given another sequence of label probabilities
+(with pauses included), marginalizing over all possible combination of
+pause lengths.
+It's used for training speech-to-text models on unaligned training data.
+
+'Most implementations of CTC compute the marginal logprob
+and its gradients combining a forward and a backward pass.
+However, as mentioned in the original paper, we should only
+need the last step of the forward pass to compute the marginal
+probability, and Dex's autodiff should produce the backward
+pass automatically.  That makes this code much shorter than
+most implementations.
+
+
+def interleave (blank:v) (labels: m=>v) : (m & (Fin 2))=>v =
+  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
+  -- then flattening the pairs back into a single-index table.
+  pairs = for i. [labels.i, blank]
+  for (i, j). pairs.i.j
+
+def prepend (first: v) (seq: m=>v) : (Unit|m)=>v =
```

Comment:
It's cool that this sum-type index set worked! I think this is the first example of it being used in anger.

Reply:
I liked that the index set can be both iterated over in a flat manner, but that the type of the index set "remembers" where the different elements came from.